### DISCUSSION TITLE
TDS Temporary Stability Simulation Problem in Different Boot States

### DISCUSSION THREAD

**Discussion by Crazy-Bike** [CURENT/andes#528] [Q&A] [answered]
Hello, I am using Andes simulation to simulate the frequency changes of the system under load disturbances in the IEEE-39 system example. Under the same disturbance, I set different generator connection states to simulate different startup methods. But I found that under certain disturbances (with a magnitude of 20% of the total load), TDS simulation will crash when all units are turned on, while TDS simulation can proceed normally when some units are turned on. This indicates that the system can withstand disturbances of 20% load. However, compared to common sense, this is contrary because when all units are turned on, their inertia is more abundant and their primary frequency regulation ability is stronger, but the simulation process is not like this.
I understand it as caused by power imbalance, but I found that power flow calculation can proceed normally, and TDS simulation should theoretically be able to proceed normally based on this. Could you please provide me with an answer.
Also, I would like to ask how the frequency spike that occurs at the moment of disturbance during the simulation process is formed?
Thank you very much！！
[ieee39_alter_shutdown.zip](https://github.com/CURENT/andes/files/14944745/ieee39_alter_shutdown.zip)
![image](https://github.com/CURENT/andes/assets/130566375/6c54e8be-8e36-4f13-baaa-571a7d6f80fd)
![image](https://github.com/CURENT/andes/assets/130566375/83ddf307-e8b3-46c1-b95f-e2d73fc8804f)

**Comment by Crazy-Bike** [comment]
![image](https://github.com/CURENT/andes/assets/130566375/3a6ed4c4-f168-4a19-9532-77d10bed54d7)
This is the instantaneous spike that appears in the frequency curve at the moment of disturbance during the simulation, and it is not known what the reason is

**Comment by cuihantao** [comment, marked as answer]
> This indicates that the system can withstand disturbances of 20% load. However, compared to common sense, this is contrary because when all units are turned on, their inertia is more abundant and their primary frequency regulation ability is stronger, but the simulation process is not like this.

I'm not sure about the cause, but I'm with you on common sense. When a dynamic generator does not take the place of a static one, the static PV generator remains to impose the fixed P and V. 

I'll look into this to rule out software bugs.

About the spike in frequency: all the frequencies obtained by BusFreq are estimates, since the network is based on steady-state phasor. You can adjust the time-constant for the washout filter to smooth it out. Alternatively, you can fetch the rotor speed at the generators as the frequency, but that's limited to the buses with generators. Milano's Frequency Divider provides a formula to distribute generator frequency to non-generator buses. It has yet to be implemented in ANDES.

**Reply by Crazy-Bike** [reply]
Thank you for your answer, Mr. Cui! I think the issue with the boot mode and TDS may be caused by the PV node. Because when changing the connection state u of the GENRU model, the connection state u of the PV nodes connected by GENRU did not change accordingly, resulting in normal results when calculating power flow based on PV nodes; But an error occurred during TDS.

**Comment by jinningwang** [comment]
As a supplement regarding the spike, here is a relevant discussion that shows how the parameters can impact the measured variables, https://github.com/CURENT/andes/discussions/476

**Reply by Crazy-Bike** [reply]
Thank you for your answer. I think it's very helpful. Wishing you a happy life!

**Comment by cuihantao** [comment]
The PV node should have been turned off automatically, if the GENROU model is specified in the case file and has u=1. This is done in the initialization process by calling v_numeric of GENROU.

If you turn off GENROU by custom code, PV will not turn off.

Did the problem get solved?

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Reply by Crazy-Bike** [reply]
Thank you for your concern about my issue, Mr. Cui! I tried to control the connection status of the PV module by adding an Alter module. However, in Andes, Alter achieves state changes by identifying the device's idx, such as "ss.PV. alter ('u ',' f {idx} ', 0)". In the Excel file of the existing case, the idx of the PV module is in the form of numbers "1, 2, 3", etc. However, when adding Alter according to the idx form of the PV module in Excel, it cannot be recognized. 
![image](https://github.com/CURENT/andes/assets/130566375/56c96179-39b4-42b2-93cf-aa4718466f69)
I tried changing the idx name in the Excel file, such as "PV_1", so that Alter could recognize it, but more problems arose when I run TDS.
![image](https://github.com/CURENT/andes/assets/130566375/6a9bddbe-3461-48be-8818-e6d98df215c9)
![image](https://github.com/CURENT/andes/assets/130566375/7c4c11e4-e76b-4302-8bba-5adef6da1155)

In the Andes document, idx is the unique identification number of the device, but in existing cases, I found that the idx of the PV module conflicts with the idx of the BUS module, which are represented only in Arabic numerals. However, no similar problem occurred during Andes operation, and I do not yet know the cause of the formation.
![image](https://github.com/CURENT/andes/assets/130566375/cefb8175-3927-4717-abf0-38b5f31a21b8)

**Reply by jinningwang** [reply]
1. Sometimes `idx=2` run into error, but `idx='2'`. It might be caused by the input type but I didn't test yet.
1. idx is unique for devices in a same `group`. Bus and PV don't share the same group.

**Reply by cuihantao** [reply]
> Sometimes idx=2 run into error, but idx='2'. It might be caused by the input type but I didn't test yet.
You can determine the type by checking the corresponding `idx.v` of the device to tell whether it's a string or an integer.

If it's ingested as a string, it's probably because in your excel file, the column format is set to Text instead of Number.

**Reply by cuihantao** [reply]
@Crazy-Bike  Just check the corresponding `ss.PV.idx.v` to figure out the type

**Reply by jinningwang** [reply]
Actually, I have a question about the `idx` type. Is it a good idea to force the `IdxParam` type to be string?
Pros: more robust and consistent data type
Cons: I don't see any

**Reply by cuihantao** [reply]
It's okay but we don't gain much by doing so. Someone may come back and ask to support numerical indices. I think it's trivial enough to let it go. Just check the type first.

**Reply by cuihantao** [reply]
Can you share the test file with issues? @Crazy-Bike

**Reply by Crazy-Bike** [reply]
This is the test file I used and the corresponding case，hoping to help solve the question.
[ieee39.xlsx](https://github.com/CURENT/andes/files/14968202/ieee39.xlsx)
[PVtest.zip](https://github.com/CURENT/andes/files/14968213/PVtest.zip)

**Reply by cuihantao** [reply]
Thanks, but I was looking for one that has "Alter".

**Comment by jinningwang** [comment]
Check follow documentation for more details about replacing a static generator with a dynamic generator, https://docs.andes.app/en/latest/groupdoc/StaticGen.html#staticgen

@cuihantao Hantao, is it allowed to change the device connection status `u` during TDS through `Model.alter()` or `Model.set()`? Another discussion also mentioned this issue but I'm not sure, https://github.com/CURENT/andes/discussions/481

**Reply by cuihantao** [reply]
> is it allowed to change the device connection status u during TDS through Model.alter() or Model.set()?

This is allowable. But if you have a GENROU with initially `u=0`, and you try to turn it on during the simulation, it will take some caution. When turning a GENROU on during the simulation, one needs to manually turn off the corresponding PV or Slack. Otherwise, GENROU and the corresponding PV will stay on at the same time.

Maybe we need to provide an interface to handle turning devices on and off, instead of allowing the user to write directly to `u`. We can probably overload the `set()` function for models, such as GENROU, that need to perform extra steps during a turn-on.

**Reply by cuihantao** [reply]
https://github.com/CURENT/andes/discussions/481 may be a separate issue. The system had two slack generators. One of them needs to be turned off upon connection.
